packages/@ember/-internals/glimmer/lib/modifiers/internal.ts

```diff
@@ -0,0 +1,107 @@
+import { Owner, setOwner } from '@ember/-internals/owner';
+import { guidFor } from '@ember/-internals/utils';
+import { assert } from '@ember/debug';
+import {
+  CapturedArguments,
+  Destroyable,
+  InternalModifierManager as ModifierManager,
+  VMArguments,
+} from '@glimmer/interfaces';
+import { valueForRef } from '@glimmer/reference';
+import { registerDestructor, setModifierManager } from '@glimmer/runtime';
+import { SimpleElement } from '@simple-dom/interface';
+
+export default class InternalModifier {
+  // Factory interface
+  static create(): never {
+    throw assert('Use constructor instead of create');
+  }
+
+  static get class(): typeof InternalModifier {
+    return this;
+  }
+
+  static get fullName(): string {
+    return this.name;
+  }
+
+  static get normalizedName(): string {
+    return this.name;
+  }
+
+  constructor(
+    protected owner: Owner,
+    protected readonly element: Element,
+    protected readonly args: CapturedArguments
+  ) {
+    setOwner(this, owner);
+  }
+
+  install(): void {}
+
+  remove(): void {}
+
+  protected positional(index: number): unknown {
+    let ref = this.args.positional[index];
+    return ref ? valueForRef(ref) : undefined;
+  }
+
+  protected named(key: string): unknown {
+    let ref = this.args.named[key];
+    return ref ? valueForRef(ref) : undefined;
+  }
+
+  toString(): string {
+    return `<${this.constructor.toString()}:${guidFor(this)}>`;
+  }
+}
+
+class InternalModifierState implements Destroyable {
+  constructor(readonly name: string, readonly instance: InternalModifier) {}
+}
+
+class InternalModifierManager
+  implements ModifierManager<InternalModifierState, typeof InternalModifier> {
+  constructor(private readonly owner: Owner) {}
+
+  create(
+    element: SimpleElement,
+    factory: typeof InternalModifier,
+    args: VMArguments
+  ): InternalModifierState {
+    assert('element must be an HTMLElement', element instanceof HTMLElement);
```

Comment:
It was a bit strange that this was a _SimpleElement_ and requires a cast, since I thought modifiers only run in the browser. Also wasn't sure if this is the canonical way to cast it these days.